common/src/main/java/net/opentsdb/threadpools/UserAwareThreadPoolExecutor.java

```diff
@@ -0,0 +1,352 @@
+          "Percentage of active threads to consider before applying the reschedule. "
+              + "Can be null to use the default.");
+    }
+
+    if (!tsdb.getConfig().hasProperty(getConfigKey(MAX_THREAD_PER_USER_PCT))) {
+      tsdb.getConfig().register(getConfigKey(MAX_THREAD_PER_USER_PCT), 50, false,
+          "Percentage of active threads per user to consider before applying the reschedule when the total "
+              + "percentage of active thread in the executor falls below the "
+              + THREAD_THRESHOLD_PCT + " threshold. " + "Can be null to use the default.");
+    }
+
+    int threadThresholdPct = tsdb.getConfig().getInt(getConfigKey(THREAD_THRESHOLD_PCT));
+    if (threadThresholdPct > 100) {
+      return Deferred.fromError(
+          new IllegalStateException(THREAD_THRESHOLD_PCT + " value should be less than 100"));
+    }
+    int threadThresholdPerUserPct = tsdb.getConfig().getInt(getConfigKey(THREAD_THRESHOLD_PCT));
+    if (threadThresholdPerUserPct > 100) {
+      return Deferred.fromError(
+          new IllegalStateException(MAX_THREAD_PER_USER_PCT + " value should be less than 100"));
+    }
+
+    Integer maxSize = tsdb.getConfig().getInt(getConfigKey(QUEUE_MAX_SIZE));
+    corePoolSize = tsdb.getConfig().getInt(getConfigKey(CORE_THREAD_POOL_SIZE));
+    maxTPoolSize = tsdb.getConfig().getInt(getConfigKey(MAX_THREAD_POOL_SIZE));
+    threadThresholdCnt = maxTPoolSize * (threadThresholdPct / 100);
+    threadThresholdPerUserCnt = maxTPoolSize * (threadThresholdPerUserPct / 100);
+
+    LOG.info(
+        "Initializing new UserAwareThreadPoolExecutor with max queue size of {} with core threads {} and a "
+            + "maximum pool size of {} with threshold Thread count of {}.",
+        maxSize, corePoolSize, maxTPoolSize, threadThresholdPct);
+
+    workQueue = new LinkedBlockingQueue<Runnable>(maxSize);
+
+    // TODO: better alternative?
+    // If the worker queue is full, we run the task in the thread that calls execute(task)
+    RejectedExecutionHandler inPlaceRunTaskHandler = new RejectedExecutionHandler() {
+
+      @Override
+      public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+        if (!executor.isShutdown()) {
+          r.run();
+        }
+
+      }
+    };
+    this.threadPool = new ThreadPoolExecutor(corePoolSize, maxTPoolSize, 1L, TimeUnit.SECONDS,
+        workQueue, inPlaceRunTaskHandler);
+
+    LOG.info("Initialized new UserAwareThreadPoolExecutor with {} threads", corePoolSize);
+    return Deferred.fromResult(null);
+  }
+
+  class QCRunnableWrapper implements Runnable {
+    final Runnable r;
+    final QueryContext qctx;
+
+    public QCRunnableWrapper(Runnable r, QueryContext qctx) {
+      this.r = r;
+      this.qctx = qctx;
+    }
+
+    @Override
+    public void run() {
+
+      if (qctx == null) {
+        r.run();
+        return;
+      }
+
+      final String user = this.qctx.authState().getUser();
```

Comment:
Check for nulls here. If they are null then use the string `Unkown` and we'll throttle on that.

Reply:
Oh,  you think we will have authstate but not user?

Reply:
yeah if auth is disabled for any reason we can get null users. There's a default user in that case.